File: answer/206/model_answer.cpp
/*
 * @lc app=leetcode id=206 lang=cpp
 *
 * [206] Reverse Linked List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode *prev = nullptr, *next = nullptr;
        while (head) {
            next = head->next; //次のptrの値を保持
            head->next = prev; //headのnextが指すptrを逆にする
            prev = head; // prevを先にすすめる
            head = next; // 保持していた値をreturn
        }
        return prev;
    }
};
// @lc code=end

File: answer/142/model_answer.go
/*
 * @lc app=leetcode id=142 lang=golang
 *
 * [142] Linked List Cycle II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func detectCycle(head *ListNode) *ListNode {
    if head == nil || head.Next == nil {
		return nil
	}
	fast := head
	slow := head
	for fast != nil && fast.Next != nil {
        slow = slow.Next
		fast = fast.Next.Next
        if fast == slow {
			break
		}
	}
	if fast != slow {
		return nil
	}
	slow = head
	for fast != slow {
		fast = fast.Next
		slow = slow.Next
	}
	return slow
}

// @lc code=end

